samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java

```diff
@@ -152,32 +165,47 @@ public ApplicationStatus status(StreamApplication streamApp) {
     return appStatus;
   }
 
+  @Override
+  public void waitForFinish() {
+    try {
+      shutdownLatch.await();
+    } catch (Throwable t) {
+      log.error("Wait is interrupted by exception", t);
+      throw new SamzaException(t);
+    }
+  }
+
   /**
-   * Create the Coordination needed by the application runner.
-   * @return {@link CoordinationUtils}
-   * @throws Exception exception
+   * Create the {@link CoordinationUtils} needed by the application runner.
+   * @return an instance of {@link CoordinationUtils}
+   * @throws Exception exceptions when creating the instance, such as {@link ClassNotFoundException} and
+   * {@link InstantiationException}
    */
-  /* package private */ CoordinationUtils getCoordinationUtils() throws Exception {
-    ApplicationConfig appConfig = new ApplicationConfig(config);
-    String clazz = appConfig.getCoordinationServiceFactoryClass();
-    if (clazz != null) {
-      CoordinationServiceFactory factory = ClassLoaderHelper.fromClassName(clazz);
-      return factory.getCoordinationService(appConfig.getGlobalAppId(), uid, config);
+  /* package private */ CoordinationUtils createCoordinationUtils() throws Exception {
+    String jobCoordinatorFactoryClassName = config.get(JobCoordinatorConfig.JOB_COORDINATOR_FACTORY, "");
+
+    // TODO: we will need a better way to package the configs with application runner
+    if (ZkJobCoordinatorFactory.class.getName().equals(jobCoordinatorFactoryClassName)) {
+      ApplicationConfig appConfig = new ApplicationConfig(config);
+      return new ZkCoordinationServiceFactory().getCoordinationService(appConfig.getGlobalAppId(), uid, config);
     } else {
       return null;
     }
   }
 
   /**
-   * Create intermediate streams.
-   * @param intStreams intermediate {@link StreamSpec}s
-   * @throws Exception exception
+   * Create intermediate streams using {@link org.apache.samza.execution.StreamManager}.
+   * If {@link CoordinationUtils} is provided, this function will first invoke leader election, Pand the leader
+   * will create the streams. All the runner processes will wait on the latch that is released after the leader finishes
+   * stream creation.
+   * @param intStreams list of intermediate {@link StreamSpec}s
+   * @throws Exception exception for latch timeout
    */
   /* package private */ void createStreams(List<StreamSpec> intStreams) throws Exception {
     if (!intStreams.isEmpty()) {
-      if (coordination != null) {
-        Latch initLatch = coordination.getLatch(1, LATCH_INIT);
-        coordination.getLeaderElector().tryBecomeLeader(() -> {
+      if (coordinationUtils != null) {
+        Latch initLatch = coordinationUtils.getLatch(1, INIT_LATCH_ID);
```

Comment:
@xinyuiscool I wasn't referring to the method name. I just wanted you to change the local variable name here as distributedLatch or sharedLatch. Can we still do that here and follow-up with refactoring and renaming in SAMZA-1262? It is difficult for me to read the code, otherwise.